text/2020-12-25-selectors.md

```diff
@@ -0,0 +1,242 @@
+selected again by "Mode/Value", then we get the final target of chaos injecting.
+
+I think we should split these two-part into different interfaces.
+
+There are several usage for the selector:
+
+- Select certain pods with pod-name.
+- List pods with Kubernetes native label/field selector from target
+  namespace(s), then filter with namespace, annotations, and pod-phase.
+- Check a pod(resource) is matches the selector.
+
+## Detailed design
+
+### SelectSpec
+
+The struct of CRD will not change.
+
+It will add another method for `SelectSpec`:
+
+```go
+type SelectSpec interface {
+    GetSelector() v1alpha1.SelectorSpec
+    GetMode() v1alpha1.PodMode
+    GetValue() string
+    ResourceType() interface{}
+}
+```
+
+It will return an empty struct with the same type of resources that will be
+selected. For example, currently, all implementation should return an empty
+struct `v1.Pod`, which means all chaos will be injected into **Pod**. **This
+method should be implemented by code generation.**
+
+> We already have requirement for injecting chaos into other resources, like
+> **Service** or some Fake/Mirrored Resource for multi-cluster.
+
+Since we will select more types of resources excepted Pod, so `v1alpha1.PodMode`
+should be changed to another name, like `ChooseMode` or something else.
+
+### Selector
+
+We will define an interface called `Selector` that could "select" expected
+resource from Kubernetes.
+
+> It still depends on Kubernetes. :) We will not leave Kubernetes in shorten
+> future.
+
+```go
+interface Selector {
+    Select() ([]interface{}, error)
+    CouldResolve() []interface{}
+}
+
+interface Filter{
+    Filter(origin []interface{}) ([]interface{}, error)
+    CouldResolve() []interface{}
+}
+```
+
+> As golang do not contain features like generic type, method
+> `Selector#Select()` just return `[]interface{}` directly. And a method
+> `CouldResolve()` will return an array of the empty struct with the same type
+> represents the resources which this Selector/Filter will return.
+
+Why CouldResolve return an array of interface{} not a single interface{}?
+
+Based on Kubernetes, it's easy to implement a shared selector that could resolve
+selecting on any types of resources. Because label/namespace and other things
+are out of CRD.
+
+Let's talk about implementations:
+
+#### Implementations of Selector
+
+##### CertainResourceSelector
+
+It select resources from Kubernetes by certain names, and it contains a
+Kubernetes client and trying to fetch resources by name one-by-one.
+
+##### ResourceFetcher
+
+It also contains Kubernetes client, and select resources from Kubernetes by
+Kubernetes native list API with label/field selectors.
+
+### Filters
+
+```go
+struct NamespaceFilter {
+    AllowedNamespaces []string
+}
+
+func (it *NamespaceFilter) Filter(origin []interface{}) ([]interface,error) {
+    // your codes here
+}
+```
+
+### SelectorChain
```

Comment:
I prefer to get a single filter, which could be a combination of multiple filter. e.g. `AND` filter:

```go
AND(aFilter, bFilter, OR(cFilter, dFilter, eFilter), fFilter)
```

Reply:
Cool!  When we accept more than one filter, we must apply them with AND/OR. `AND` and `OR` could create various filter logic with compositing many `Filter`s.

I will update this doc.